src/streamlit_extras/stateful_chat/__init__.py

```diff
@@ -0,0 +1,150 @@
+from __future__ import annotations
+
+import time
+from typing import TYPE_CHECKING, Any, List, Sequence, Union
+
+import streamlit as st
+from streamlit.elements.image import AtomicImage
+from streamlit.errors import StreamlitAPIException
+from typing_extensions import Literal, Required, TypedDict
+
+from streamlit_extras import streaming_write
+
+from .. import extra
+
+if TYPE_CHECKING:
+    from streamlit.delta_generator import DeltaGenerator
+
+SpecType = Union[int, Sequence[Union[int, float]]]
+
+
+class ChatMessage(TypedDict):
+    author: Required[str]
+    avatar: Required[str | AtomicImage | None]
+    content: Required[List[Any]]
+
+
+def _active_dg():
+    from streamlit.runtime.scriptrunner import get_script_run_ctx
+
+    ctx = get_script_run_ctx()
+    if ctx and len(ctx.dg_stack) > 0:
```

Comment:
What is dg_stack?

Reply:
It's the current stack of delta generators. This code is basically a replication of our internal code to get the active delta generator: 

https://github.com/streamlit/streamlit/blob/3d5dd61aef2bf649c5bfe42aad53cf1426beb094/lib/streamlit/delta_generator.py#L323

Reply:
![plotting-the-simpsons-gif](https://github.com/arnaudmiribel/streamlit-extras/assets/4040678/5046dd3a-2c51-47ec-a251-5cdb3ef205c5)